File: CBTReframe/ViewModels/SettingsViewModel.swift
import Foundation
import SwiftUI
import SwiftData
import UIKit
import LocalAuthentication

protocol ReminderScheduling {
    func requestPermission() async -> Bool
    func scheduleDailyReminder(hour: Int, minute: Int) async
    func cancelDailyReminder()
}

struct DefaultReminderScheduler: ReminderScheduling {
    func requestPermission() async -> Bool {
        await ReminderService.requestPermission()
    }

    func scheduleDailyReminder(hour: Int, minute: Int) async {
        await ReminderService.scheduleDailyReminder(hour: hour, minute: minute)
    }

    func cancelDailyReminder() {
        ReminderService.cancelDailyReminder()
    }
}

@MainActor
@Observable
final class SettingsViewModel {
    private static let modelCacheKeyPrefix = "cachedModelList."
    private let reminderScheduler: ReminderScheduling
    private var reminderSyncTask: Task<Void, Never>?

    /// 从 API 拉取并持久化后的模型；无缓存时用 `fallbackModels`。
    private var modelCache: [String: [AIModel]] = [:]

    var isRefreshingModels = false
    var modelsListError: String?

    var selectedProvider: AIProvider {
        didSet {
            UserDefaults.standard.set(selectedProvider.rawValue, forKey: "selectedProvider")
            let models = resolvedModels(for: selectedProvider)
            if !models.contains(where: { $0.id == selectedModelId }) {
                selectedModelId = selectedProvider.resolveDefaultModelId(from: models)
            }
            loadAPIKey()
            Task { await refreshModels() }
        }
    }

    var selectedModelId: String {
        didSet {
            UserDefaults.standard.set(selectedModelId, forKey: "selectedModelId")
        }
    }

    var apiKeyInput: String = ""
    var isSavingAPIKey = false
    var useFaceID: Bool {
        didSet {
            UserDefaults.standard.set(useFaceID, forKey: "useFaceID")
        }
    }
    var dailyReminderEnabled: Bool {
        didSet {
            UserDefaults.standard.set(dailyReminderEnabled, forKey: "dailyReminderEnabled")
            syncReminderSchedule()
        }
    }
    var reminderHour: Int {
        didSet {
            UserDefaults.standard.set(reminderHour, forKey: "reminderHour")
            syncReminderSchedule()
        }
    }
    var reminderMinute: Int {
        didSet {
            UserDefaults.standard.set(reminderMinute, forKey: "reminderMinute")
            syncReminderSchedule()
        }
    }
    var hasAcceptedDisclaimer: Bool {
        didSet {
            UserDefaults.standard.set(hasAcceptedDisclaimer, forKey: "hasAcceptedDisclaimer")
        }
    }

    func resolvedModels(for provider: AIProvider) -> [AIModel] {
        if provider == .local { return AIProvider.local.fallbackModels }
        if let cached = modelCache[provider.rawValue], !cached.isEmpty {
            return cached
        }
        return provider.fallbackModels
    }

    var selectedModel: AIModel {
        let list = resolvedModels(for: selectedProvider)
        return list.first { $0.id == selectedModelId }
            ?? selectedProvider.fallbackModels.first!
    }

    init(reminderScheduler: ReminderScheduling = DefaultReminderScheduler()) {
        self.reminderScheduler = reminderScheduler
        let providerRaw = UserDefaults.standard.string(forKey: "selectedProvider") ?? AIProvider.gemini.rawValue
        let provider = AIProvider(rawValue: providerRaw) ?? .local
        self.selectedProvider = provider

        let modelId = UserDefaults.standard.string(forKey: "selectedModelId") ?? ""
        self.selectedModelId = modelId.isEmpty ? provider.defaultModel.id : modelId

        self.useFaceID = UserDefaults.standard.bool(forKey: "useFaceID")
        self.dailyReminderEnabled = UserDefaults.standard.bool(forKey: "dailyReminderEnabled")
        self.reminderHour = UserDefaults.standard.object(forKey: "reminderHour") as? Int ?? 21
        self.reminderMinute = UserDefaults.standard.object(forKey: "reminderMinute") as? Int ?? 0
        self.hasAcceptedDisclaimer = UserDefaults.standard.bool(forKey: "hasAcceptedDisclaimer")

        loadModelCacheFromDisk()
        let list = resolvedModels(for: selectedProvider)
        if !list.contains(where: { $0.id == selectedModelId }) {
            selectedModelId = provider.resolveDefaultModelId(from: list)
        }

        loadAPIKey()
        Task { await refreshModels() }
    }

    private func loadModelCacheFromDisk() {
        for p in AIProvider.allCases where p.requiresAPIKey {
            let key = Self.modelCacheKeyPrefix + p.rawValue
            guard let data = UserDefaults.standard.data(forKey: key),
                  let models = try? JSONDecoder().decode([AIModel].self, from: data) else { continue }
            modelCache[p.rawValue] = models
        }
    }

    private func persistModelCache(_ models: [AIModel], for provider: AIProvider) {
        modelCache[provider.rawValue] = models
        if let data = try? JSONEncoder().encode(models) {
            UserDefaults.standard.set(data, forKey: Self.modelCacheKeyPrefix + provider.rawValue)
        }
    }

    func initializeOnboarding() async throws {
        let provider = selectedProvider
        let key = apiKeyInput.trimmingCharacters(in: .whitespacesAndNewlines)
        
        if provider.requiresAPIKey {
            if key.isEmpty {
                throw NSError(domain: "Settings", code: 1, userInfo: [NSLocalizedDescriptionKey: "API Key 不能为空"])
            }
            // 保存 key 到 Keychain
            KeychainManager.shared.save(key: provider.rawValue, value: key)
        } else {
            KeychainManager.shared.delete(key: provider.rawValue)
        }
        
        // 拉取模型并验证
        if provider.requiresAPIKey {
            let models = try await AIModelListService.fetchModels(provider: provider, apiKey: key)
            if models.isEmpty {
                throw NSError(domain: "Settings", code: 2, userInfo: [NSLocalizedDescriptionKey: "未能获取到任何可用模型，请检查配置"])
            }
            persistModelCache(models, for: provider)
            if !models.contains(where: { $0.id == selectedModelId }) {
                selectedModelId = provider.resolveDefaultModelId(from: models)
            }
        }
        
        hasAcceptedDisclaimer = true
    }

    func loadAPIKey() {
        apiKeyInput = KeychainManager.shared.load(key: selectedProvider.rawValue) ?? ""
    }

    /// 从服务商拉取最新模型列表（需已在 Keychain 中保存有效 Key）。
    func refreshModels() async {
        guard selectedProvider.requiresAPIKey else {
            modelsListError = nil
            return
        }
        let key = KeychainManager.shared.load(key: selectedProvider.rawValue) ?? ""
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            modelsListError = nil
            return
        }

        isRefreshingModels = true
        modelsListError = nil
        defer { isRefreshingModels = false }

        do {
            let models = try await AIModelListService.fetchModels(provider: selectedProvider, apiKey: trimmed)
            if !models.isEmpty {
                persistModelCache(models, for: selectedProvider)
                if !models.contains(where: { $0.id == selectedModelId }) {
                    selectedModelId = selectedProvider.resolveDefaultModelId(from: models)
                }
            }
        } catch {
            modelsListError = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
        }
    }

    func saveAPIKey() {
        let trimmed = apiKeyInput.trimmingCharacters(in: .whitespacesAndNewlines)
        let providerKey = selectedProvider.rawValue
        guard !isSavingAPIKey else { return }
        isSavingAPIKey = true
        Task {
            await Task.detached {
                if trimmed.isEmpty {
                    KeychainManager.shared.delete(key: providerKey)
                } else {
                    KeychainManager.shared.save(key: providerKey, value: trimmed)
                }
            }.value
            isSavingAPIKey = false
            let feedback = UINotificationFeedbackGenerator()
            feedback.notificationOccurred(.success)
            await refreshModels()
        }
    }

    func clearAllData(modelContext: ModelContext) {
        KeychainManager.shared.deleteAll()
        apiKeyInput = ""

        if let historyEntries = try? modelContext.fetch(FetchDescriptor<HistoryEntry>()) {
            for entry in historyEntries {
                modelContext.delete(entry)
            }
        }

        if let thoughtEntries = try? modelContext.fetch(FetchDescriptor<ThoughtEntry>()) {
            for entry in thoughtEntries {
                modelContext.delete(entry)
            }
        }
        if let moodEntries = try? modelContext.fetch(FetchDescriptor<MoodCheckIn>()) {
            for entry in moodEntries {
                modelContext.delete(entry)
            }
        }

        try? modelContext.save()

        for p in AIProvider.allCases where p.requiresAPIKey {
            UserDefaults.standard.removeObject(forKey: Self.modelCacheKeyPrefix + p.rawValue)
        }
        modelCache.removeAll()

        let domain = Bundle.main.bundleIdentifier ?? "com.cbt.reframe"
        UserDefaults.standard.removePersistentDomain(forName: domain)
        selectedProvider = .local
        selectedModelId = AIProvider.local.defaultModel.id
        useFaceID = false
        dailyReminderEnabled = false
        reminderHour = 21
        reminderMinute = 0
        hasAcceptedDisclaimer = false
    }

    var hasAPIKey: Bool {
        guard selectedProvider.requiresAPIKey else { return true }
        let key = KeychainManager.shared.load(key: selectedProvider.rawValue) ?? ""
        return !key.isEmpty
    }

    func authenticateWithFaceID(reason: String) async -> Bool {
        let context = LAContext()
        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            return false
        }
        do {
            return try await context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: reason
            )
        } catch {
            return false
        }
    }

    private func syncReminderSchedule() {
        reminderSyncTask?.cancel()
        let enabled = dailyReminderEnabled
        let hour = reminderHour
        let minute = reminderMinute
        let scheduler = reminderScheduler

        reminderSyncTask = Task {
            if enabled {
                _ = await scheduler.requestPermission()
                guard !Task.isCancelled else { return }
                await scheduler.scheduleDailyReminder(hour: hour, minute: minute)
            } else {
                scheduler.cancelDailyReminder()
            }
        }
    }
}

File: shared/src/commonMain/kotlin/com/henryliu/cbtreframe/shared/SettingsViewModel.kt
package com.henryliu.cbtreframe.shared

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class SettingsUiState(
    val selectedProvider: AIProvider = AIProvider.GEMINI,
    val selectedModelId: String = "",
    val apiKeyInput: String = "",
    val isSavingAPIKey: Boolean = false,
    val isRefreshingModels: Boolean = false,
    val modelsListError: String? = null,
    val useFaceID: Boolean = false,
    val dailyReminderEnabled: Boolean = false,
    val reminderHour: Int = 21,
    val reminderMinute: Int = 0,
    val hasAcceptedDisclaimer: Boolean = false,
    val resolvedModels: List<AIModel> = emptyList(),
    val hasAPIKey: Boolean = false,
    val modelInvalidated: Boolean = false,
    val invalidatedModelName: String? = null,
) {
    val selectedModel: AIModel
        get() = resolvedModels.firstOrNull { it.modelName == selectedModelId }
            ?: resolvedModels.firstOrNull()
            ?: AIModel(AIProvider.GEMINI, "gemini-flash-latest", "Gemini Flash Latest")
}

interface KeychainProvider {
    fun load(key: String): String?
    fun save(key: String, value: String)
    fun delete(key: String)
    fun deleteAll()
}

interface ReminderScheduler {
    suspend fun requestPermission(): Boolean
    suspend fun scheduleDailyReminder(hour: Int, minute: Int)
    fun cancelDailyReminder()
}

class SettingsViewModel(
    private val settingsManager: SettingsManager,
    private val keychainProvider: KeychainProvider,
    private val biometricAuthProvider: BiometricAuthProvider,
    private val reminderScheduler: ReminderScheduler,
    private val modelFetcher: ModelFetcher,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
) {
    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    private var reminderSyncJob: Job? = null

    init {
        // Load persisted state
        val provider = settingsManager.getSelectedProvider()
        val modelId = settingsManager.getSelectedModelId().let {
            it.ifEmpty { provider.defaultModelId() }
        }

        val models = loadResolvedModels(provider)

        val savedModelId = if (models.any { m -> m.modelName == modelId }) modelId
            else provider.defaultModelId()

        _uiState.value = _uiState.value.copy(
            selectedProvider = provider,
            selectedModelId = savedModelId,
            apiKeyInput = keychainProvider.load(provider.name) ?: "",
            useFaceID = settingsManager.getUseFaceID(),
            dailyReminderEnabled = settingsManager.getDailyReminderEnabled(),
            reminderHour = settingsManager.getReminderHour(),
            reminderMinute = settingsManager.getReminderMinute(),
            hasAcceptedDisclaimer = settingsManager.getHasAcceptedDisclaimer(),
            resolvedModels = models,
            hasAPIKey = computeHasAPIKey(provider),
        )

        scope.launch { refreshModels() }
    }

    fun clear() {
        scope.cancel()
    }

    // ── Provider selection ─────────────────────────────────────────────

    fun selectProvider(provider: AIProvider) {
        println("SELECT_PROVIDER provider=$provider")
        val models = loadResolvedModels(provider)
        val modelId = if (models.any { it.modelName == _uiState.value.selectedModelId })
            _uiState.value.selectedModelId
        else
            provider.defaultModelId()

        settingsManager.setSelectedProvider(provider)
        settingsManager.setSelectedModelId(modelId)

        _uiState.value = _uiState.value.copy(
            selectedProvider = provider,
            selectedModelId = modelId,
            apiKeyInput = keychainProvider.load(provider.name) ?: "",
            resolvedModels = models,
            hasAPIKey = computeHasAPIKey(provider),
        )

        println("SELECT_PROVIDER launching refreshModels")
        scope.launch { refreshModels() }
    }

    fun selectModel(modelId: String) {
        settingsManager.setSelectedModelId(modelId)
        _uiState.value = _uiState.value.copy(selectedModelId = modelId)
    }

    // ── API Key ────────────────────────────────────────────────────────

    fun updateApiKeyInput(value: String) {
        _uiState.value = _uiState.value.copy(apiKeyInput = value)
    }

    fun saveAPIKey() {
        val trimmed = _uiState.value.apiKeyInput.trim()
        val provider = _uiState.value.selectedProvider
        val providerKey = provider.name

        if (_uiState.value.isSavingAPIKey) return

        _uiState.value = _uiState.value.copy(isSavingAPIKey = true)

        scope.launch {
            withContext(Dispatchers.Default) {
                if (trimmed.isEmpty()) {
                    keychainProvider.delete(providerKey)
                } else {
                    keychainProvider.save(providerKey, trimmed)
                }
            }
            _uiState.value = _uiState.value.copy(
                isSavingAPIKey = false,
                hasAPIKey = computeHasAPIKey(provider),
            )
            refreshModels()
        }
    }

    suspend fun validateAndFetchModels(provider: AIProvider, apiKey: String): Result<Unit> {
        return try {
            println("INIT-1 start")
            // 1. Save provider and API key
            println("INIT: saving provider")
            settingsManager.setSelectedProvider(provider)
            println("INIT-2 provider saved")
            val providerKey = provider.name
            if (provider.requiresApiKey()) {
                if (apiKey.isBlank()) {
                    return Result.failure(Exception("API Key 不能为空"))
                }
                println("INIT: saving api key")
                withContext(Dispatchers.Default) {
                    keychainProvider.save(providerKey, apiKey.trim())
                }
            } else {
                println("INIT: saving api key")
                withContext(Dispatchers.Default) {
                    keychainProvider.delete(providerKey)
                }
            }
            println("INIT-3 key saved")

            // Sync ViewModel UI State locally
            _uiState.value = _uiState.value.copy(
                selectedProvider = provider,
                apiKeyInput = apiKey,
                hasAPIKey = computeHasAPIKey(provider)
            )
            println("INIT-4 ui updated")

            // 2. Fetch models & cache
            if (provider.requiresApiKey()) {
                println("INIT-5 before fetchModels")
                val models = withTimeout(15000) {
                    modelFetcher.fetchModels(provider, apiKey.trim())
                }
                println("INIT-6 after fetchModels")
                if (models.isEmpty()) {
                    return Result.failure(Exception("未能获取到任何可用模型，请检查配置"))
                }
                println("INIT: caching models")
                settingsManager.setCachedModels(provider, models)
                val currentModelId = _uiState.value.selectedModelId
                val newModelId = if (models.any { it.modelName == currentModelId }) currentModelId
                    else provider.resolveDefaultModelId(models)

                _uiState.value = _uiState.value.copy(
                    resolvedModels = models,
                    selectedModelId = newModelId,
                    modelsListError = null
                )
            } else {
                val fallback = provider.fallbackModels()
                val newModelId = provider.resolveDefaultModelId(fallback)
                _uiState.value = _uiState.value.copy(
                    resolvedModels = fallback,
                    selectedModelId = newModelId,
                    modelsListError = null
                )
            }

            println("INIT-7 success")
            Result.success(Unit)
        } catch (e: Exception) {
            println("INIT: validateAndFetchModels failed: ${e::class.simpleName}: ${e.message}")
            e.printStackTrace()
            _uiState.value = _uiState.value.copy(
                modelsListError = e.message ?: "Unknown error"
            )
            Result.failure(e)
        }
    }

    fun completeOnboarding(selectedModelId: String) {
        settingsManager.setSelectedModelId(selectedModelId)
        settingsManager.setHasAcceptedDisclaimer(true)
        _uiState.value = _uiState.value.copy(
            selectedModelId = selectedModelId,
            hasAcceptedDisclaimer = true
        )
    }

    // ── Model refresh ──────────────────────────────────────────────────

    suspend fun refreshModels() {
        println("REFRESH: enter")
        val provider = _uiState.value.selectedProvider
        println("REFRESH: provider=$provider")
        println("REFRESH: requiresApiKey=${provider.requiresApiKey()}")
        if (!provider.requiresApiKey()) {
            val fallback = provider.fallbackModels()
            _uiState.value = _uiState.value.copy(
                resolvedModels = fallback,
                modelsListError = null,
            )
            return
        }

        val key = (keychainProvider.load(provider.name) ?: "").trim()
        println("REFRESH: keyLoaded length=${key.length}")
        if (key.isEmpty()) {
            _uiState.value = _uiState.value.copy(modelsListError = null)
            return
        }

        _uiState.value = _uiState.value.copy(
            isRefreshingModels = true,
            modelsListError = null,
        )

        try {
            println("REFRESH: before fetchModels")
            val models = modelFetcher.fetchModels(provider, key)
            println("REFRESH: after fetchModels count=${models.size}")
            if (models.isNotEmpty()) {
                settingsManager.setCachedModels(provider, models)
                val currentModelId = _uiState.value.selectedModelId
                val wasInvalidated = currentModelId.isNotEmpty() && models.none { it.modelName == currentModelId }

                val newModelId = if (!wasInvalidated && models.any { it.modelName == currentModelId }) {
                    currentModelId
                } else {
                    provider.resolveDefaultModelId(models)
                }

                settingsManager.setSelectedModelId(newModelId)

                _uiState.value = _uiState.value.copy(
                    resolvedModels = models,
                    selectedModelId = newModelId,
                    modelInvalidated = wasInvalidated,
                    invalidatedModelName = if (wasInvalidated) currentModelId else null
                )
            }
            println("REFRESH: success path finished")
        } catch (e: Exception) {
            println("REFRESH: exception=${e::class.simpleName}: ${e.message}")
            e.printStackTrace()
            _uiState.value = _uiState.value.copy(
                modelsListError = e.message ?: "Unknown error",
            )
        } finally {
            println("REFRESH: finally")
            _uiState.value = _uiState.value.copy(isRefreshingModels = false)
        }
    }

    fun dismissModelInvalidationBanner() {
        _uiState.value = _uiState.value.copy(modelInvalidated = false, invalidatedModelName = null)
    }

    // ── Toggles ────────────────────────────────────────────────────────

    fun setUseFaceID(enabled: Boolean) {
        settingsManager.setUseFaceID(enabled)
        _uiState.value = _uiState.value.copy(useFaceID = enabled)
    }

    fun setDailyReminderEnabled(enabled: Boolean) {
        settingsManager.setDailyReminderEnabled(enabled)
        _uiState.value = _uiState.value.copy(dailyReminderEnabled = enabled)
        syncReminderSchedule()
    }

    fun setReminderHour(hour: Int) {
        settingsManager.setReminderHour(hour)
        _uiState.value = _uiState.value.copy(reminderHour = hour)
        syncReminderSchedule()
    }

    fun setReminderMinute(minute: Int) {
        settingsManager.setReminderMinute(minute)
        _uiState.value = _uiState.value.copy(reminderMinute = minute)
        syncReminderSchedule()
    }

    fun setHasAcceptedDisclaimer(accepted: Boolean) {
        settingsManager.setHasAcceptedDisclaimer(accepted)
        _uiState.value = _uiState.value.copy(hasAcceptedDisclaimer = accepted)
    }

    // ── Face ID auth ───────────────────────────────────────────────────

    suspend fun authenticateWithBiometrics(reason: String): Boolean {
        return biometricAuthProvider.authenticate(reason, "")
    }

    val canUseBiometrics: Boolean
        get() = biometricAuthProvider.canAuthenticate()

    // ── Clear all ──────────────────────────────────────────────────────

    fun clearAllData(onClearDatabase: suspend () -> Unit) {
        keychainProvider.deleteAll()
        settingsManager.clearAll()
        _uiState.value = SettingsUiState(
            selectedProvider = AIProvider.LOCAL,
            selectedModelId = AIProvider.LOCAL.defaultModelId(),
            resolvedModels = AIProvider.LOCAL.fallbackModels(),
        )
        scope.launch { onClearDatabase() }
    }

    // ── Private helpers ────────────────────────────────────────────────

    private fun loadResolvedModels(provider: AIProvider): List<AIModel> {
        if (provider == AIProvider.LOCAL) return provider.fallbackModels()
        val cached = settingsManager.getCachedModels(provider)
        return if (cached.isNullOrEmpty()) provider.fallbackModels() else cached
    }

    private fun computeHasAPIKey(provider: AIProvider): Boolean {
        if (!provider.requiresApiKey()) return true
        val key = keychainProvider.load(provider.name) ?: ""
        return key.isNotEmpty()
    }

    private fun syncReminderSchedule() {
        reminderSyncJob?.cancel()
        val enabled = _uiState.value.dailyReminderEnabled
        val hour = _uiState.value.reminderHour
        val minute = _uiState.value.reminderMinute

        reminderSyncJob = scope.launch {
            if (enabled) {
                val granted = reminderScheduler.requestPermission()
                if (!granted || !isActive) return@launch
                reminderScheduler.scheduleDailyReminder(hour, minute)
            } else {
                reminderScheduler.cancelDailyReminder()
            }
        }
    }
}

// ── Extension helpers ─────────────────────────────────────────────────────

fun AIProvider.requiresApiKey(): Boolean = this != AIProvider.LOCAL

fun AIProvider.defaultModelId(): String {
    return fallbackModels().first().modelName
}

fun AIProvider.resolveDefaultModelId(models: List<AIModel>): String {
    val firstId = models.firstOrNull()?.modelName ?: defaultModelId()
    if (this != AIProvider.GEMINI) return firstId
    
    val priority = listOf(
        "gemini-flash-latest",
        "gemini-2.5-flash",
        "gemini-2.0-flash",
        "gemini-2.0-flash-lite",
        "gemini-1.5-flash"
    )
    val idSet = models.map { it.modelName }.toSet()
    for (id in priority) {
        if (idSet.contains(id)) return id
    }
    return firstId
}

fun AIProvider.fallbackModels(): List<AIModel> {
    return when (this) {
        AIProvider.OPENAI -> listOf(
            AIModel(this, "gpt-4.1", "GPT-4.1"),
            AIModel(this, "gpt-4.1-mini", "GPT-4.1 Mini"),
            AIModel(this, "gpt-4.1-nano", "GPT-4.1 Nano"),
            AIModel(this, "gpt-4o", "GPT-4o"),
            AIModel(this, "gpt-4o-mini", "GPT-4o Mini")
        )
        AIProvider.ANTHROPIC -> listOf(
            AIModel(this, "claude-sonnet-4-20250514", "Claude Sonnet 4"),
            AIModel(this, "claude-3-5-haiku-20241022", "Claude 3.5 Haiku")
        )
        AIProvider.DEEPSEEK -> listOf(
            AIModel(this, "deepseek-chat", "DeepSeek Chat"),
            AIModel(this, "deepseek-reasoner", "DeepSeek Reasoner")
        )
        AIProvider.GEMINI -> listOf(
            AIModel(this, "gemini-flash-latest", "Gemini Flash Latest"),
            AIModel(this, "gemini-2.5-flash", "Gemini 2.5 Flash"),
            AIModel(this, "gemini-2.0-flash", "Gemini 2.0 Flash"),
            AIModel(this, "gemini-2.0-flash-lite", "Gemini 2.0 Flash-Lite"),
            AIModel(this, "gemini-1.5-flash", "Gemini 1.5 Flash"),
            AIModel(this, "gemini-1.5-pro", "Gemini 1.5 Pro")
        )
        AIProvider.KIMI -> listOf(
            AIModel(this, "moonshot-v1-8k", "Moonshot v1 8K"),
            AIModel(this, "moonshot-v1-32k", "Moonshot v1 32K"),
            AIModel(this, "kimi-k2-turbo-preview", "Kimi K2 Turbo"),
            AIModel(this, "kimi-k2-thinking-preview", "Kimi K2 Thinking")
        )
        AIProvider.LOCAL -> listOf(
            AIModel(this, "local", "内置分析")
        )
    }
}

// ── Model fetching interface ───────────────────────────────────────────────

interface ModelFetcher {
    suspend fun fetchModels(provider: AIProvider, apiKey: String): List<AIModel>
}
